org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/editor/quickfix/XtextGrammarQuickfixProvider.java

```diff
@@ -146,32 +148,53 @@ public void fixUnresolvedRule(final Issue issue, IssueResolutionAcceptor accepto
 				"Create rule '" + ruleName + "'", //
 				"Create rule '" + ruleName + "'", //
 				NULL_QUICKFIX_IMAGE, //
-				createNewRule(ruleName, false));
+				createNewRule(ruleName, null));
 		createLinkingIssueResolutions(issue, acceptor);
 	}
 
+	@Fix(XtextLinkingDiagnosticMessageProvider.UNRESOLVED_RULE)
 	@Fix(XtextLinkingDiagnosticMessageProvider.UNRESOLVED_TERMINAL_RULE)
 	public void fixUnresolvedTerminalRule(final Issue issue, IssueResolutionAcceptor acceptor) {
-		final String ruleName = issue.getData()[0];
+		String ruleName = issue.getData()[0];
+		String ruleType = "terminal ";
 		acceptor.accept(issue, //
 				"Create terminal '" + ruleName + "'", // 
 				"Create terminal '" + ruleName + "'", //
 				NULL_QUICKFIX_IMAGE, //
-				createNewRule(ruleName, false));
-		createLinkingIssueResolutions(issue, acceptor);
+				createNewRule(ruleName, ruleType));
+
+		Optional<IssueResolution> iResolution = acceptor.getIssueResolutions().stream().filter(ir -> {
```

Comment:
i am not sure if i like the after effect filtering
shouldnt 
createLinkingIssueResolutions
be modified to ctreate resolutions for that matching methods rule type only

Reply:
@szarnekow what do you think?

Reply:
I'd be in favor of adding another `@Fix(UNRESOLVED_TERMINAL_RULE)` and call `createLinkingIssueResolutions()` from there to make sure it is only invoked once for that issue code.